### DISCUSSION TITLE
When to Arc state and when to not Arc state

### DISCUSSION THREAD

**Discussion by askasp** [tokio-rs/axum#2710] [Q&A]
### Summary

I struggle a bit with understanding the implications of Arc in Axum state handler. I would really appreciate it if anyone has the time to clarify what is best practice here.
Specifically, i’m unsure if its enough to have the entire struct in an arc, or i also should have the db client in a separate one.
Alt1:
```rust
struct AppState{
  prisma_client: PrismaClient
  stripe_client: StripeClient
}

and then create one like this

let app_state = Arc::new(AppState{...})

let app = Router::new()
          .route("/user/:id", put(user::update))
          .route("/stripe", post(stripe::create))
          .with_state(app_state.clone()),
   
```    
Or if i should wrap every part in each own Arc?
Alt2:
```rust
struct AppState{
  prisma_client: Arc<PrismaClient>
  stripe_client: Arc<StripeClient>
}
```
How would axum actually handle the later case? I’ve understood it so that it automatically clones the struct in each handler, is that correct?
Will i in the later case be able to do
```rust
pub async fn stripe(State(app_state): State<AppState>) {
  let query = app_state.prisma_client.query  
```
or do i need to explicitly clone the prisma_client here?

Generally i wonder what is the best practice, and the implications of each approach! 

### axum version

0.74

**Comment by wangziling** [comment]
Hi~ I thought that these two cases are all ok.

According to the existed practice about the `crates.io`.
https://github.com/rust-lang/crates.io/blob/e3a14fbc4beda762784daf59a946c18b26a8dc90/src/app.rs#L22

It only wrapped the `config` and `storage` with `Arc`, and I only found the `.clone()` of they two here: https://github.com/rust-lang/crates.io/blob/e3a14fbc4beda762784daf59a946c18b26a8dc90/src/tests/util/test_app.rs#L277.

Hmmm...
In my opinion, I thought that:

```rs
// let app_state = AppState {
//     config: Arc<Config>,
//     client: Arc<Client>,
//}
```

If we want to share the `config` or `client` in multiple places without clone themselves' metrics or data totally, then we need to wrap them with `Arc` as well. 😃
